ntoskrnl/mm/ARM3/sysldr.c

```diff
@@ -2964,7 +2964,52 @@ MmLoadSystemImage(IN PUNICODE_STRING FileName,
     PrefixName = *FileName;
 
     /* Check if we have a prefix */
-    if (NamePrefix) DPRINT1("Prefixed images are not yet supported!\n");
+    if (NamePrefix)
+    {
+        /* Check if "directory + prefix" is too long for the string */
+        Status = RtlUShortAdd(BaseDirectory.Length,
```

Comment:
Neither here nor below do you reserve place for a null terminator.

Reply:
The prefixed name is a counted string, it therefore doesn't necessarily need a null terminator.

Reply:
I wasn't clear about this, originally assumed a string structure with a length property won't strictly need a null terminator. I checked `RtlInitUnicodeString` and it did reserve the place for null terminator. This may need more changes than just a few lines...

Reply:
RtlInitUnicodeString does that because the C-string passed in order to initialize the UNICODE_STRING has a null-terminator.

Reply:
According to MSDN https://learn.microsoft.com/en-us/windows/win32/api/ntdef/ns-ntdef-_unicode_string ,

> If the string is null-terminated, Length does not include the trailing null character.

So the `UNICODE_STRING` doesn't force the string to be null-terminated, nor does the `Length` cares about it. Am I wrong about RtlInitUnicodeString or is the implementation inaccurate?

Reply:
That's correct. And RtlInitUnicodeString is perfectly fine. This does not conflict with those two remarks:
> If the string is null-terminated, Length does not include the trailing null character.

and
> The MaximumLength is used to **indicate the length of Buffer** so that if the string is passed to a conversion routine such as RtlAnsiStringToUnicodeString the returned string does not exceed the buffer size.

Reply:
Oh yeah it's MaximumLength not Length *facepalm*

Reply:
it doesn't _need_ it, but i think it's good practice.